### PR TITLE
[NXCM-5131] Fix itemMaxAge settings for non-maven repositories

### PR DESCRIPTION
**Pull request by nabcos**
This pull reverts the initial change for NXCM-5131, because it affected behavior of non-maven-repositories inheriting from AbstractProxyRepository. instead of the implicit default of itemMaxAge (1440 minutes), 0 would be used instead.

With this change, the repository resource sets itemMaxAge for repositories that do not have the maven-specific artifact/metadataMaxAge.

### CODE REVIEW

**Comment by ifedorenko** [comment]
Do we need a UT for "implicit default value of 1440 minutes is used"?

**Comment by nabcos** [comment]
After some more code reading, the previous fix does not break anything for "real life" situations:

The NuGet IT that fails creates the proxy repository with _unset_ itemMaxAge, which would previously fall back to the implicit default for "no external configuration". 

If you create a repository in the UI, that value will always be set explicitely, so there is no behavior change visible for end users. The itemMaxAge is actually used in NuGet repositories, and can be set and saved in the UI. (It will not be read back without this change to the RepositoryPlexusResource, though.)

File: nexus-client-core/src/main/java/org/sonatype/nexus/client/core/subsystem/repository/BaseRepository.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package org.sonatype.nexus.client.core.subsystem.repository;
+
+/**
+ * Base class for hosted/proxy repositories.
+ *
+ * @since 2.4
+ */
+interface BaseRepository<T extends Repository, S extends RepositoryStatus>
+    extends Repository<T, S>
```
Comment (jdillon) [inline nexus-client-core/src/main/java/org/sonatype/nexus/client/core/subsystem/repository/BaseRepository.java:21]: is this package private on purpose?

File: plugins/restlet1x/nexus-restlet1x-testsuite/pom.xml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    <dependencies>
+      <dependency>
+        <groupId>org.sonatype.nexus.plugins</groupId>
+        <artifactId>nexus-restlet1x-testsupport-plugin</artifactId>
+        <version>${project.version}</version>
+        <classifier>bundle</classifier>
+        <type>nexus-plugin</type>
+      </dependency>
+    </dependencies>
+  </dependencyManagement>
```
Comment (jdillon) [inline plugins/restlet1x/nexus-restlet1x-testsuite/pom.xml]: this should be defined in nexus-restlet1x-parent

File: plugins/restlet1x/nexus-restlet1x-testsuite/pom.xml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    </dependency>
+
+    <dependency>
+      <groupId>${it.nexus.bundle.groupId}</groupId>
+      <artifactId>${it.nexus.bundle.artifactId}</artifactId>
+      <version>${it.nexus.bundle.version}</version>
+      <scope>test</scope>
+      <type>zip</type>
+      <classifier>bundle</classifier>
+    </dependency>
```
Comment (jdillon) [inline plugins/restlet1x/nexus-restlet1x-testsuite/pom.xml:40]: Why is this here, this should already be defined

File: nexus-client-core/src/main/java/org/sonatype/nexus/client/core/subsystem/repository/BaseRepository.java (hunk 1 above)
Comment (nabcos) [inline nexus-client-core/src/main/java/org/sonatype/nexus/client/core/subsystem/repository/BaseRepository.java:21]: Yes, I did not see much value in exposing that technical DRY-step as public API, but the intermediate step is needed IIRC because virtual and group repositories don't have that property.

File: nexus-client-core/src/main/java/org/sonatype/nexus/client/core/subsystem/repository/BaseRepository.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+package org.sonatype.nexus.client.core.subsystem.repository;
+
+/**
+ * Base class for hosted/proxy repositories.
+ *
+ * @since 2.5
+ */
+interface BaseRepository<T extends Repository, S extends RepositoryStatus>
```
Comment (ifedorenko) [inline nexus-client-core/src/main/java/org/sonatype/nexus/client/core/subsystem/repository/BaseRepository.java:20]: What's the point of this interface? Why 'browsable' cannot be not part of Repository?

Comment (nabcos) [inline nexus-client-core/src/main/java/org/sonatype/nexus/client/core/subsystem/repository/BaseRepository.java:20]: Virtual and group repositories don't have browsable.

File: nexus-client-core/src/main/java/org/sonatype/nexus/client/core/subsystem/repository/Repository.java [hunk 5]
```diff
@@ -83,4 +83,9 @@
      */
     T remove();
 
+    /**
+     * @return {@code true} if the repository is exposed, {@code false} otherwise.
+     * @since 2.4
```
Comment (ifedorenko) [inline nexus-client-core/src/main/java/org/sonatype/nexus/client/core/subsystem/repository/Repository.java]: wrong @since version (and in few other places)

File: plugins/restlet1x/nexus-restlet1x-plugin/src/main/java/org/sonatype/nexus/rest/repositories/AbstractRepositoryPlexusResource.java [hunk 6]
```diff
@@ -392,22 +392,26 @@ public RepositoryProxyResource getRepositoryProxyRestModel( ProxyRepository repo
             // This is a total hack to be able to retrieve this data from a non core repo if available
```
Comment (ifedorenko) [inline plugins/restlet1x/nexus-restlet1x-plugin/src/main/java/org/sonatype/nexus/rest/repositories/AbstractRepositoryPlexusResource.java:392]: This is just sick. :-( I know we had this marvel of software design for awhile now, but is there a chance to fix this properly, i.e. without resorting to reflection, and without rewriting half of nexus?

File: nexus-client-core/src/main/java/org/sonatype/nexus/client/core/subsystem/repository/BaseRepository.java (hunk 4 above)
Comment (ifedorenko) [inline nexus-client-core/src/main/java/org/sonatype/nexus/client/core/subsystem/repository/BaseRepository.java:20]: How is this related to artifact maxage?

File: plugins/restlet1x/nexus-restlet1x-plugin/src/main/java/org/sonatype/nexus/rest/repositories/AbstractRepositoryPlexusResource.java (hunk 6 above)
Comment (cstamas) [inline plugins/restlet1x/nexus-restlet1x-plugin/src/main/java/org/sonatype/nexus/rest/repositories/AbstractRepositoryPlexusResource.java:392]: Yes, just make UI use Templates for repo creation. In short, changes I was proposing long time ago:

1) expose existing templates over REST
2) Repo creation UI should offer them... so, instead as today (select provider "maven2" and repository policy "Released", plus implicitly selected kind when you pressed Add Hosted / Proxy / ...., just make UI show the list of existing (filtered) templates.
3) a bit more work needed on templates

So, if we want to stick with Add Hosted / Proxy ... buttons, we could filter templates for HostedRepository, ProxyRepository etc, and then instead of setting "maven2" and "Release" (from corresponding drop downs), user would choose "Maven2 (hosted, release)" entry (which maps to org.sonatype.nexus.templates.repository.maven.Maven2HostedRepositoryTemplate.Maven2HostedRepositoryTemplate) and UI would just send back the used template ID.

Still, even if this would not require rewriting half of nexus, templates (org.sonatype.nexus.templates.repository package) and this resource would need some love, as then there still the need to "apply" the values to the configuration sent by UI to the template before template.create() is invoked (which actually creates live repo instance, and does everything needed (adds it to config, saves config etc).

File: nexus-client-core/src/main/java/org/sonatype/nexus/client/core/subsystem/repository/BaseRepository.java (hunk 4 above)
Comment (nabcos) [inline nexus-client-core/src/main/java/org/sonatype/nexus/client/core/subsystem/repository/BaseRepository.java:20]: The IT checks default values for unset properties on repo creation, not only for itemMaxAge.

**Comment by nabcos** [comment]
https://bamboo.zion.sonatype.com/browse/NXO-OSSF0-S9-1

**Comment by jdillon** [comment]
I'm not sure what to do about this, I don't know the core api well enough to say if this is good or if there is something better.  I don't think this makes it worse however (its already pretty bad), and if this fixes the problem then we probably should merge it.
